theano/tensor/basic.py

```diff
@@ -3340,43 +3340,236 @@ def transpose(x, axes=None):
     return ret
 
 
-def batched_dot(x, y):
+class BatchedDot(Op):
     """
-    This function computes the dot product between the two tensors, by
-    iterating over the first dimension using scan.
+    Dot but Batched
 
-    Parameters
-    ----------
-    x : tensor
-        A Tensor with sizes e.g.: for  3D (dim1, dim3, dim2).
-    y : tensor
-        A Tensor with sizes e.g.: for 3D (dim1, dim2, dim4).
+    WRITEME
+    Computes the dot product of two variables. For two matrices, this is
+    equivalent to matrix multiplication. For two vectors, this is the inner
+    product.
 
-    Returns
-    -------
-    tensor
-        A tensor of size e.g. if it is 3D: (dim1, dim3, dim4).
+    Notes
+    -----
+    Matrix-matrix products are sometimes optimized to Dot22 or Gemm ops
+    (see tensor.blas).
+    Vector-vector products are sometimes optimized to Ger or CGer (see
+    tensor.blas).
+    Matrix-vector products are sometimes optimized to Gemv, CGemv (see
+    tensor.blas).
+
+    """
+    __props__ = ()
+
+    def make_node(self, *inputs):
+        inputs = list(map(as_tensor_variable, inputs))
+
+        if len(inputs) != 2:
+            raise TypeError(
+                'theano.tensor.BatchedDot: 2 arguments required, %d given ' %
+                len(inputs))
+        if inputs[0].ndim not in (2, 3):
+            raise TypeError(
+                'theano.tensor.BatchedDot: input 0 (0-indexed) must have ndim'
+                ' of 2 or 3, %d given. Consider calling '
+                'theano.tensor.batched_dot instead.' % inputs[0].ndim)
+        if inputs[1].ndim not in (2, 3):
+            raise TypeError(
+                'theano.tensor.BatchedDot: input 1 (0-indexed) must have ndim'
+                'of 2 or 3, %d given. Consider calling '
+                'theano.tensor.batched_dot instead.' % inputs[1].ndim)
+
+        dtype = scal.upcast(*[input.type.dtype for input in inputs])
+        broadcastable = (inputs[0].type.broadcastable[:-1] +
+                         inputs[1].type.broadcastable[2:])
```

Comment:
I hadn't thought about this, but now that I have, that doesn't seem right. The way I understand it, a dimension should be broadcastable only if it was broadcastable in both inputs. Otherwise the output will have a definite (possibly >1) size in that dimension.